### DISCUSSION TITLE
Exportable Human-Readable Library talk

### DISCUSSION THREAD

**Discussion by Qronikarz** [TagStudioDev/TagStudio#250] [General]
Opening topic because I read it in commit history and I worked on something similar for QMV and TagsResearch.

So, tl;dr I tried designing human readable export/import tag data for TagsResearch, but couldn't settle on anything. So maybe there can be discussion with more people how to approach this. Bear in mind that this export should be useful for as many programs as possible and shouldn't use the same assumptions like single programs - for example QMV internally replaces whitespace with underscore (" " -> "_"), but other programs don't have to do that and so the export should use whitespaces.

First attempt with same line name and tags. https is for bookmarks, file.png is the file, /directory is for folders. The last two can have whitespaces in them so you need to take care of that. Also last two are relative path so the export can be cross-platform. The importing program should ask for the accurate location.
~~~
"https://www.ndoesnte.com" : "tag1" "tag2" "tag3" "tag 4" "tag cz"

"file.png" : "tag1" "tag2" "tag3" "tag 4" "tag cz"

"/directory" : "tag1" "tag2" "tag3" "tag 4" "tag cz"

"/dir/directory/file.png" : "tag1" "tag2" "tag3" "tag 4" "tag cz"

"/dir/directory/file name.png" : "tag1" "tag2" "tag3" "tag 4" "tag cz"

"/dir/directory name" : "tag1" "tag2" "tag3" "tag 4" "tag cz"

"custom item" : "tag1" "tag2" "tag3" "tag 4" "tag cz"
~~~
Wasn't very readable for my taste.

Second attempt with name in separate line
~~~
https://www.ndoesnte.com
"tag1" "tag2" "tag3" "tag 4" "tag cz"

file.png
"tag1" "tag2" "tag3" "tag 4" "tag cz"

/directory
"tag1" "tag2" "tag3" "tag 4" "tag cz"

/dir/directory/file.png
"tag1" "tag2" "tag3" "tag 4" "tag cz"

/dir/directory/file name.png 
"tag1" "tag2" "tag3" "tag 4" "tag cz"

/dir/directory name
"tag1" "tag2" "tag3" "tag 4" "tag cz"
~~~
Better because you don't need to use "" to surround file names with whitespaces

Last attempt:
~~~
ste programname
https://www.ndoesnte.com (tag1, tag2, tag3, tag 4, tag cz)
file.png (tag1, tag2, tag3, tag 4, tag cz)
/directory (tag1, tag2, tag3, tag 4, tag cz)
/dir/directory/file.png (tag1, tag2, tag3, tag 4, tag cz)
/dir/directory/file name.png (tag1, tag2, tag3, tag 4, tag cz)
/dir/directory name (tag1, tag2, tag3, tag 4, tag cz)
custom item (tag1, tag2, tag3, tag 4, tag cz)
~~~
(ste - simple tag export, programname - name of program that exported this text)
I actually used it in the newest QMV version 0.3 to provide tags when the breakage happens, to not lose your tags and it looks like that:
~~~
ste QuoMediaView query:()
example_qmv_base/files/VanGogh-starry_night.jpg (picture, jpeg, sky, clouds, stars, moon, 1889, Van Gogh)
example_qmv_base/files/tapetadwaosiem.png (picture, png, unknown year, Qronikarz)
example_qmv_base/files/Sturmtief_Paul_2017-06-22.webm.360p.vp9.webm (video, webm, audio, grass field, sky, trees, unknown year, rain, wind)
example_qmv_base/files/Starry_Night_Over_the_Rhone.jpg (picture, jpeg, river, sky, stars, 1888, Van Gogh)
example_qmv_base/files/stanczyk.jpg (picture, jpeg, 1862, Stańczyk, 34 yo, chair, Jan Matejko, black eyes, jester suit)
~~~
My files don't have any spaces, but it doesn't matter right now since I didn't make the import function for it yet and also added the query so you can know from what search the export came.

One problem is it's still not exactly human readable for me. And mind you, these are just normal, non-nested tags without any relations. When it comes to TagStudio tags it will be harder.
